File: cppLeetcode/leetcodeM/1319_NumberOfOperationsToMakeNetworkConnected.cpp
/*
 1319. Number of Operations to Make Network Connected
 
 There are n computers numbered from 0 to n-1 connected by ethernet cables connections forming a network where connections[i] = [a, b]
 represents a connection between computers a and b. Any computer can reach any other computer directly or indirectly through the network.

 Given an initial computer network connections. You can extract certain cables between two directly connected computers,
 and place them between any pair of disconnected computers to make them directly connected. Return the minimum number of times
 you need to do this in order to make all the computers connected. If it's not possible, return -1.

  

 Example 1:



 Input: n = 4, connections = [[0,1],[0,2],[1,2]]
 Output: 1
 Explanation: Remove cable between computer 1 and 2 and place between computers 1 and 3.
 Example 2:



 Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2],[1,3]]
 Output: 2
 Example 3:

 Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2]]
 Output: -1
 Explanation: There are not enough cables.
 Example 4:

 Input: n = 5, connections = [[0,1],[0,2],[3,4],[2,3]]
 Output: 0
  

 Constraints:

 1 <= n <= 10^5
 1 <= connections.length <= min(n*(n-1)/2, 10^5)
 connections[i].length == 2
 0 <= connections[i][0], connections[i][1] < n
 connections[i][0] != connections[i][1]
 There are no repeated connections.
 No two computers are connected by more than one cable.
 
 */
#include <vector>
#include <unordered_map>

using std::unordered_map;
using std::vector;

class NumberOfOperationsMakeNetworkConnected {

    /*
        1319.Number-of-Operations-to-Make-Network-Connected
        首先想到的是，属于同一组互联的多个计算机可能会有一些冗余的连接线。那么冗余多少呢？很简单，有N个计算机已经互联的话，那么只需要N-1条线就可以保证它们互联，多余的线其实都是可以拔掉的，我们可以改用它来连接不同局域网。

        接下来，我们知道，如果将M个不相连的局域网都互通，那么需要M-1条线。

        所以总体的算法是：用Union Find的方法来数一下有M个局域网；再计算每个局域网里有多少冗余的传输线并相加。如果所有冗余的传输线数目K>=M-1的话，那么就可以把所有的计算机都连接起来。
    */
     vector<int>Father;
public:
    int makeConnected(int n, vector<vector<int>>& connections) 
    {
        Father.resize(n);
        vector<int>edges(n,0);
        unordered_map<int,int>cables;
        unordered_map<int,int>elements;

        for (int i=0; i<n; i++)
            Father[i] = i;

        for (auto& edge: connections)
        {
            if (FindFather(edge[0])!=FindFather(edge[1]))
                Union(edge[0],edge[1]);
            edges[edge[0]]+=1;
            edges[edge[1]]+=1;
        }

        for (int i=0; i<n; i++)
        {
            Father[i] = FindFather(i);
            cables[Father[i]] += edges[i];
            elements[Father[i]] += 1;
        }

        int redundant = 0;
        int islands = cables.size();
        for (auto x: cables)
        {
            int root = x.first;
            redundant += cables[root]/2+1-elements[root];
        }

        if (redundant>=islands-1)
            return islands-1;
        else 
            return -1;        
    }

    int FindFather(int x)
    {
        if (Father[x]!=x)
            Father[x] = FindFather(Father[x]);
        return Father[x];
    }
    
    void Union(int x, int y)
    {
        x = Father[x];
        y = Father[y];
        if (x<y)
            Father[y] = x;
        else
            Father[x] = y;
    }


    
    struct UnionFind{
        int num;
        int numComponents;
        vector<int> parent;
    
        UnionFind(int n){
            num = n;
            numComponents = n;
            parent = vector<int>(n);
            
            for(int i = 0; i<n; i++)
                 parent[i] = i;
        }
        
        int size(){return num;}
        
        int components(){return numComponents;}
        
        int find(int p){
            while(p != parent[p]){
                parent[p] = parent[parent[p]];
                p = parent[p];
            }
            return p;
        }
        
        void unionn(int p, int q){
            int rootp = find(p), rootq = find(q);
            if(rootp == rootq) return;
            
            if(rootp < rootq)
                parent[rootq] = rootp;
            else
                parent[rootp] = rootq;
                
            numComponents--;
        }
    };
    
public:

    int makeConnected(int n, vector<vector<int>>& connections) {
        
        int redundant = 0, required = 0;
        UnionFind uf(n);
        
        for(int i = 0; i<connections.size(); i++) {
            if (uf.find(connections[i][0]) == uf.find(connections[i][1]))
                redundant++;
            else
                uf.unionn(connections[i][0], connections[i][1]);
        }
        
        return uf.components()-1 <= redundant ? uf.components()-1 : -1;
    }

public:
    
    struct UnionFind{
        int num;
        int numComponents;
        vector<int> rank;
        vector<int> parent;
    
        UnionFind(int n){
            num = n;
            numComponents = n;
            rank = vector<int>(n, 0);
            parent = vector<int>(n);
            
            for(int i = 0; i<n; i++)
                 parent[i] = i;
        }
        
        int size(){return num;}
        
        int components(){return numComponents;}
        
        int find(int p){
            while(p != parent[p]){
                parent[p] = parent[parent[p]];
                p = parent[p];
            }
            return p;
        }
        
        void unionn(int p, int q){
            int rootp = find(p), rootq = find(q);
            if(rootp == rootq) return;
            
            if(rank[rootp] < rank[rootq])
                parent[rootp] = rootq;
            else{
                parent[rootq] = rootp;
                if(rank[rootp] == rank[rootq]){rank[rootp]++;}
            }
            numComponents--;
        }
        
        int connected(int p, int q){
            return find(p) == find(q);
        }
    };
    
public:

    int doit_disjoint(int n, vector<vector<int>>& connections) {
        
        int redundant = 0, required = 0;
        UnionFind uf(n);
        
        for(int i = 0; i<connections.size(); i++) {
            if (uf.find(connections[i][0]) == uf.find(connections[i][1]))
                redundant++;
            else
                uf.unionn(connections[i][0], connections[i][1]);
        }
        
        return uf.components()-1 <= redundant ? uf.components()-1 : -1;
    }

};

File: PythonLeetcode/leetcodeM/1319_NumberOfOperationsToMakeNetworkConnected.py
"""
1319. Number of Operations to Make Network Connected

There are n computers numbered from 0 to n-1 connected by ethernet cables connections forming a network where connections[i] = [a, b]
represents a connection between computers a and b. Any computer can reach any other computer directly or indirectly through the network.

Given an initial computer network connections. You can extract certain cables between two directly connected computers,
and place them between any pair of disconnected computers to make them directly connected. Return the minimum number of
times you need to do this in order to make all the computers connected. If it's not possible, return -1.



Example 1:



Input: n = 4, connections = [[0,1],[0,2],[1,2]]
Output: 1
Explanation: Remove cable between computer 1 and 2 and place between computers 1 and 3.
Example 2:



Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2],[1,3]]
Output: 2
Example 3:

Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2]]
Output: -1
Explanation: There are not enough cables.
Example 4:

Input: n = 5, connections = [[0,1],[0,2],[3,4],[2,3]]
Output: 0


Constraints:

1 <= n <= 10^5
1 <= connections.length <= min(n*(n-1)/2, 10^5)
connections[i].length == 2
0 <= connections[i][0], connections[i][1] < n
connections[i][0] != connections[i][1]
There are no repeated connections.
No two computers are connected by more than one cable.


"""


class MakeNetworkConnected:

    class UnionFind:

        def __init__(self, n):
            self._components = n
            self._rank = [0] * n
            self._parent = [i for i in range(n)]

        def find(self, i):
            while i != self._parent[i]:
                self._parent[i] = self._parent[self._parent[i]]
                i = self._parent[i]
            return i

        def union(self, a, b):
            pa, pb = self.find(a), self.find(b)
            if pa == pb:
                return

            if self._rank[pa] > self._rank[pb]:
                self._parent[pb] = pa
                self._rank[pa] += 1
            else:
                self._parent[pa] = pb
                self._rank[pb] += 1

            self._components -= 1

        def component(self):
            return self._components

    def doit_disjoint(self, n: int, connections: list) -> int:

        uf = MakeNetworkConnected.UnionFind(n)
        redundent = 0

        for c, e in connections:
            if uf.find(c) == uf.find(e):
                print((c, e))
                redundent += 1
            else:
                uf.union(c, e)

        print(redundent)
        print(uf.component() - 1)
        return uf.component() - 1 if (uf.component() - 1) <= redundent else -1

    def doit_wrong(self, n: int, connections: list) -> int:

        redundent = 0
        visited = set()

        for s, e in connections:

            if s in visited and e in visited:
                redundent += 1
                print((s, e))

            visited.add(s)
            visited.add(e)

        print(visited)
        print(len(visited))
        print(redundent)
        return (n - len(visited)) if redundent >= n - len(visited) else -1


if __name__ == '__main__':

    # MakeNetworkConnected().doit_disjoint(n = 4, connections = [[0,1],[0,2],[1,2]])

    # MakeNetworkConnected().doit_disjoint(n = 6, connections = [[0,1],[0,2],[0,3],[1,2],[1,3]])

    # MakeNetworkConnected().doit_disjoint(n = 6, connections = [[0,1],[0,2],[0,3],[1,2]])

    # MakeNetworkConnected().doit_disjoint(n = 5, connections = [[0,1],[0,2],[3,4],[2,3]])

    MakeNetworkConnected().doit_disjoint(100, [[17, 51], [33, 83], [53, 62], [25, 34], [35, 90], [29, 41], [14, 53],
                                               [40, 84], [41, 64], [13, 68], [44, 85], [57, 58], [50, 74], [20, 69],
                                               [15, 62], [25, 88], [4, 56], [37, 39], [30, 62], [69, 79], [33, 85],
                                               [24, 83], [35, 77], [2, 73], [6, 28], [46, 98], [11, 82], [29, 72],
                                               [67, 71], [12, 49], [42, 56], [56, 65], [40, 70], [24, 64], [29, 51],
                                               [20, 27], [45, 88], [58, 92], [60, 99], [33, 46], [19, 69], [33, 89],
                                               [54, 82], [16, 50], [35, 73], [19, 45], [19, 72], [1, 79], [27, 80],
                                               [22, 41], [52, 61], [50, 85], [27, 45], [4, 84], [11, 96], [0, 99],
                                               [29, 94], [9, 19], [66, 99], [20, 39], [16, 85], [12, 27], [16, 67],
                                               [61, 80], [67, 83], [16, 17], [24, 27], [16, 25], [41, 79], [51, 95],
                                               [46, 47], [27, 51], [31, 44], [0, 69], [61, 63], [33, 95], [17, 88],
                                               [70, 87], [40, 42], [21, 42], [67, 77], [33, 65], [3, 25], [39, 83],
                                               [34, 40], [15, 79], [30, 90], [58, 95], [45, 56], [37, 48], [24, 91],
                                               [31, 93], [83, 90], [17, 86], [61, 65], [15, 48], [34, 56], [12, 26],
                                               [39, 98], [1, 48], [21, 76], [72, 96], [30, 69], [46, 80], [6, 29],
                                               [29, 81], [22, 77], [85, 90], [79, 83], [6, 26], [33, 57], [3, 65],
                                               [63, 84], [77, 94], [26, 90], [64, 77], [0, 3], [27, 97], [66, 89],
                                               [18, 77], [27, 43]])

    MakeNetworkConnected().doit(100, [[17, 51], [33, 83], [53, 62], [25, 34], [35, 90], [29, 41], [14, 53],
                                               [40, 84], [41, 64], [13, 68], [44, 85], [57, 58], [50, 74], [20, 69],
                                               [15, 62], [25, 88], [4, 56], [37, 39], [30, 62], [69, 79], [33, 85],
                                               [24, 83], [35, 77], [2, 73], [6, 28], [46, 98], [11, 82], [29, 72],
                                               [67, 71], [12, 49], [42, 56], [56, 65], [40, 70], [24, 64], [29, 51],
                                               [20, 27], [45, 88], [58, 92], [60, 99], [33, 46], [19, 69], [33, 89],
                                               [54, 82], [16, 50], [35, 73], [19, 45], [19, 72], [1, 79], [27, 80],
                                               [22, 41], [52, 61], [50, 85], [27, 45], [4, 84], [11, 96], [0, 99],
                                               [29, 94], [9, 19], [66, 99], [20, 39], [16, 85], [12, 27], [16, 67],
                                               [61, 80], [67, 83], [16, 17], [24, 27], [16, 25], [41, 79], [51, 95],
                                               [46, 47], [27, 51], [31, 44], [0, 69], [61, 63], [33, 95], [17, 88],
                                               [70, 87], [40, 42], [21, 42], [67, 77], [33, 65], [3, 25], [39, 83],
                                               [34, 40], [15, 79], [30, 90], [58, 95], [45, 56], [37, 48], [24, 91],
                                               [31, 93], [83, 90], [17, 86], [61, 65], [15, 48], [34, 56], [12, 26],
                                               [39, 98], [1, 48], [21, 76], [72, 96], [30, 69], [46, 80], [6, 29],
                                               [29, 81], [22, 77], [85, 90], [79, 83], [6, 26], [33, 57], [3, 65],
                                               [63, 84], [77, 94], [26, 90], [64, 77], [0, 3], [27, 97], [66, 89],
                                               [18, 77], [27, 43]])